presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -266,17 +297,33 @@ public static boolean stIsValid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
     @SqlNullable
     public static Slice invalidReason(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
     {
-        Geometry geometry = JtsGeometrySerde.deserialize(input);
-        if (geometry == null) {
-            return null;
-        }
+        GeometryCursor cursor = deserialize(input).getEsriGeometryCursor();
+        NonSimpleResult result = new NonSimpleResult();
+        while (true) {
+            com.esri.core.geometry.Geometry geometry = cursor.next();
+            if (geometry == null) {
+                return null;
+            }
 
-        TopologyValidationError error = new IsValidOp(geometry).getValidationError();
-        if (error == null) {
-            return null;
-        }
+            if (!OperatorSimplifyOGC.local().isSimpleOGC(geometry, null, true, result, null)) {
+                String reasonText = NON_SIMPLE_REASONS.getOrDefault(result.m_reason, result.m_reason.name());
+
+                verify(geometry instanceof MultiVertexGeometry);
```

Comment:
I meant rather to do something like `if (!(geometry instanceof MultiVertexGeometry)) return uft8Slice(reason)`